### DISCUSSION TITLE
Attachment not available for retrieval immediately?

### DISCUSSION THREAD

**Discussion by georgiosd** [ravendb/ravendb#14262] [Q&A]
This is another odd problem that's driving me crazy - I can't find anything wrong on the app level. 

Pseudo code:

app:
```
var order = await session.LoadAsync<Order>(orderId);
var receipt = GenerateReceipt();
session.Advanced.Store(order, receiptKey, receipt); // simplified
await session.StoreAsync(new SendOrderReceipt(orderId));
await session.SaveChangesAsync();
```

subscription: `from SendOrderReceipts where ProcessedOn = null include OrderId`
```
var order = await session.LoadAsync<Order>(cmd.OrderId); // INCLUDED IN SUBSCRIPTION QUERY
var attachment = await session.Advanced.Attachments.GetAsync(order, receiptKey);
```
`attachment` is null :-\

As far as I can understand the attachments feature this is not normal. 

Am I missing something or should I try to repro and file a bug report?

**Comment by grisha-kotler** [comment]
Can you create a failing test?

**Comment by ayende** [comment]
How are you opening the session?
I think that I have a good idea what is going on here.
You have something like this, right?

```
subscriptionWorker.Run(batch => 
{
       using session = documentStore.OpenSession();
       // do stuff
      session.SaveChanges();
});
```

If this is the case, then the problem is that the source node for the batch is different from the destination node for writes, which can happen when you are running in a cluster and processing a document before it replicated to all the nodes.

You can use the `batch.OpneSession()`, which will use the source node for writes.

**Reply by georgiosd** [reply]
![image](https://user-images.githubusercontent.com/1672054/172680397-073f04ed-c8a6-49c4-b16b-938a5384e358.png)

That would be an obvious problem, but sadly it's not it.

I'll see if I can create a failing test - it only fails ~1/20 times.

**Reply by ayende** [reply]
See if you can look at the traffic watch to see what the interaction looks like from the server point

**Reply by georgiosd** [reply]
@ayende Still getting this and I can't find a repro, even on the same cluster.

I enabled Operations logs for all nodes - what would I be looking for?

**Reply by ayende** [reply]
Are you reading & writing to the same node?
Operations log will not cover this, I think.

If you can reproduce even 1/20 in iterations, that is sufficient to investigate.

**Reply by georgiosd** [reply]
It happens on our production system, but I wasn't able to reproduce it with test code. 

But, I could deduce from the Sentry breadcrumbs that the subscription that stores the attachment was processed on a different node than the subscription that reads the attachment, so it would seem the only way for this to happen is if the subscription picks up the document to be processed before the attachment is committed.

Is the document and attachments committed in different steps during replication to other nodes?

**Reply by garayx** [reply]
when there is a modification to the attachment the document also get changed in same tx, so the replication of doc and the attachment should happen in the same time.

You can enable information logs on all cluster nodes, that should show the change vector that got replicated, you can share the logs us.
Also, can you verify there is no places that interfere with subscription logic I mean for example setting `ProcessedOn` to null, without adding attachment.

**Reply by georgiosd** [reply]
Yes, I can verify that. The documents that trigger the subscription (and have `ProcessedOn`) are completely separate - the attachment is on the `Order` document itself. Will forward logs to the support email shortly!

**Reply by ayende** [reply]
Are they being persisted in separate transactions?
If they are on the same transaction, they should be replicated atomically.

**Reply by georgiosd** [reply]
@ayende @garayx reviving this because we are still seeing this happen. The document and attachment are stored in the same session.

```
var messageCommand = new PostSlackMessageCommand
{
    Message = ...
};

await session.StoreAsync(messageCommand);

session.Advanced.Attachments.Store(messageCommand, attachmentName, normalizedCsvData, "text/csv");
```

This document is then picked up by a subscription - in 1/50 times, the attachment is not there yet, and I have to retry it.

Any ideas?

**Reply by garayx** [reply]
can you please describe the environment you are running on?
Are you using a cluster ?
In case yes, can you verify (either using traffic watch or information logs) that the node that processes the subscription is already got the attachment from replication (ie the processing node is `A`, but the attachment was saved to `B` and you try to load the attachment in the subscription before the attachment was added) ?

**Reply by georgiosd** [reply]
3-node cluster indeed. The issue doesn't happen every time so it's hard to replicate. 

Oren said above that the document and attachments should be replicated atomically. Am I misunderstanding what that means?

**Reply by garayx** [reply]
they should as long they were saved under same `SaveChanges()` call, can you capture traffic watch & info logs from all three nodes when the issue occurs, and send us to support email?

**Comment by georgiosd** [comment]
I am exploring this issue with support, will revert.

**Comment by georgiosd** [comment]
Is there a way to capture that programatically? It really is hard to
capture something that happens 1/10 or less.

Georgios

**Reply by garayx** [reply]
one option would be to save traffic watch logs to disk:
![image](https://github.com/ravendb/ravendb/assets/16625387/1ffa5799-f9ec-4e4e-9e0b-b204149e9129)

second one would be to use the traffic watch tool from `Raven.Debug`:


```
.\Raven.Debug(.exe) traffic --url="{SERVER_URL}" --output-path="{OUTPUT_FOLDER}" --certificate-path="{PATE_TO_PFX_FILE}" --certificate-password="{CERTIFICATE_PASSWORD}"

(example for our live-test server: .\Raven.Debug.exe traffic --url="http://live-test.ravendb.net" --output-path="C:\logs")

output-path should be created before running the tool
```

**Reply by garayx** [reply]
can also resend the previous info logs + traffic watch logs, so I can go over them again?

**Reply by georgiosd** [reply]
@garayx sadly I don't have them any more. I added a notification to catch this the next time it occurs. 

Will setting logs to Information and sending you the file logs be enough?
